src/crate/client/http.py

```diff
@@ -86,25 +86,31 @@ def super_len(o):
     return None
 
 
-class CrateJsonEncoder(json.JSONEncoder):
-    epoch_aware = datetime(1970, 1, 1, tzinfo=timezone.utc)
-    epoch_naive = datetime(1970, 1, 1)
-
-    def default(self, o):
-        if isinstance(o, (Decimal, UUID)):
-            return str(o)
-        if isinstance(o, datetime):
-            if o.tzinfo is not None:
-                delta = o - self.epoch_aware
-            else:
-                delta = o - self.epoch_naive
-            return int(
-                delta.microseconds / 1000.0
-                + (delta.seconds + delta.days * 24 * 3600) * 1000.0
-            )
-        if isinstance(o, date):
-            return calendar.timegm(o.timetuple()) * 1000
-        return json.JSONEncoder.default(self, o)
+epoch_aware = datetime(1970, 1, 1, tzinfo=timezone.utc)
+epoch_naive = datetime(1970, 1, 1)
+
+
+def cratedb_json_encoder(obj):
+    """
+    Encoder function for orjson.
+
+    https://github.com/ijl/orjson#default
+    https://github.com/ijl/orjson#opt_passthrough_datetime
+    """
+    if isinstance(obj, (Decimal, UUID)):
+        return str(obj)
+    if isinstance(obj, datetime):
+        if obj.tzinfo is not None:
+            delta = obj - epoch_aware
+        else:
+            delta = obj - epoch_naive
+        return int(
+            delta.microseconds / 1000.0
+            + (delta.seconds + delta.days * 24 * 3600) * 1000.0
+        )
+    if isinstance(obj, date):
+        return calendar.timegm(obj.timetuple()) * 1000
+    return obj
```

Comment:
In CrateDB Toolkit, I am deriving from the now-pruned traditional CrateJsonEncoder per [CrateJsonEncoderWithNumPy](https://github.com/crate/cratedb-toolkit/blob/a69a9ca3a5819f6d6fb09f941c040a6c772bed4e/cratedb_toolkit/util/sqlalchemy.py#L31-L57).

I am wondering if and how the [relevant extension that accounts for NumPy types](https://github.com/crate/cratedb-toolkit/blob/a69a9ca3a5819f6d6fb09f941c040a6c772bed4e/cratedb_toolkit/util/sqlalchemy.py#L48-L56) should be added to the core implementation instead, in order to make it more versatile, but without adding NumPy as a required dependency.

Reply:
> [...] wondering if and how the relevant extension that accounts for NumPy types should be added.

It looks like this will become a non-issue, because [orjson](https://pypi.org/project/orjson/) can handle NumPy types already:

> It serializes [dataclass](https://github.com/ijl/orjson?tab=readme-ov-file#dataclass), [datetime](https://github.com/ijl/orjson?tab=readme-ov-file#datetime), [numpy](https://github.com/ijl/orjson?tab=readme-ov-file#numpy), and [UUID](https://github.com/ijl/orjson?tab=readme-ov-file#uuid) instances natively.

Reply:
Do you think we should add relevant software tests to crate-python, which marshal a few examples of NumPy types, while the NumPy package will only become part of the **test dependencies** in crate-python?

I think any other framework naturally would do downstream integration tests on their own behalves, like sqlalchemy-cratedb is doing it, and many others, out of band of the test suite of crate-python.

However, because the relevant fragments of code in focus here are extending the core type marshalling function `cratedb_json_encoder`, it may be applicable to integrate a small fragment of NumPy testing code into the test suite itself, now that we know NumPy types will probably not need any special treatments, just to make sure about it.

Reply:
> Do you think we should add relevant software tests to crate-python, which marshal a few examples of NumPy types, while the NumPy package will only become part of the test dependencies in crate-python?

As you wrote `because [orjson](https://pypi.org/project/orjson/) can handle NumPy types already:` it seems that this feature is purely inside `orjson` so it should be tested there already. I don't see any reason to test the functionality of a dependency inside our code. Do I miss something? ;)

Reply:
All right. I just wanted to check (better safe than sorry), I am mostly bringing it up about not missing support for NumPy types.

Now, after [reading up on the docs](https://github.com/ijl/orjson?tab=readme-ov-file#numpy), I can see:

> orjson is faster than all compared libraries at serializing numpy instances. Serializing numpy data requires specifying `option=orjson.OPT_SERIALIZE_NUMPY`.

So, I guess we'd like to enable this option, if you don't have any objections.

Reply:
Enabled per 20e735fcf.

Reply:
https://github.com/crate/cratedb-toolkit/pull/349 shows that NumPy support works well, so we have been able to remove relevant monkeypatches needed before. 💯